### ISSUE TITLE
Monitoring Causes High Memory Uses/Crashes

### ISSUE THREAD

**Issue by flaccidbagel** [Bionus/imgbrd-grabber#1695]
**Steps to reproduce**

1. Enable Monitoring for any set of favorited tags.
2. Set any interval
3. Run the program

**Context**

When using the monitoring function, it appears that once the initial monitoring delay in the options is met, all configured monitors begin running on their normal interval, at the same time. For anyone with a moderate to large number of favorited tags, this leads to a very quick loss of performance within the program, to an eventual crash without a crash log being written. 

Not sure if this should be more of a suggestion to stagger the monitoring so that only one tag is being searched at a time, or a crash report due to the inevitable crash that occurs. Typically around 1.7 GB of Ram on my machine.

**System information**

- OS: Windows 10
- Grabber version: 7.1.1
- 16GB Ram

**Additional context**

Wish I had some logs to upload, but there are no crash logs being generated from this particular crash. I do have the windows event viewer log however:

Faulting application name: Grabber.exe, version: 0.0.0.0, time stamp: 0x5ce14e37
Faulting module name: Qt5Widgets.dll, version: 5.10.1.0, time stamp: 0x5a7c81e9
Exception code: 0xc000041d
Fault offset: 0x0000000000041568
Faulting process id: 0x18a0
Faulting application start time: 0x01d5183309af5a74
Faulting application path: D:\%appdata\%logs\Programs\Grabber\Grabber.exe
Faulting module path: D:\%appdata\%logs\Programs\Grabber\Qt5Widgets.dll
Report Id: 5a016f26-ab45-4b18-8769-3837163b9816
Faulting package full name: 
Faulting package-relative application ID:

**Comment by Bionus** [comment]
Okay, so there seem to be two separate issues at play here:

### Performance issue

While there is always only one monitor running at the same time, if you have multiple monitors set with the same interval, they will be loaded one after the other without any pause, and it seem to happen in the main thread, causing freezes/performance issues in the application.

A few possible solutions:
* Add an interval between each check (so if two are supposed to be run at a given point, add a 1 second "pause" between both)
* Run them all in parallel (might cause issues if multiple monitors use the same source, and might cause CPU "spikes" in the program, which I'd rather avoid)
* Run the monitor checker in a separate thread (I don't know why it's not already the case, as it should be)

I guess I'll go with simply moving the monitoring to another thread, as it should solve all issues.

### Memory usage

The high memory usage, followed by crashes, is very likely caused by a memory leak in the monitoring system, so I'll investigate and get rid of that.

**Comment by flaccidbagel** [comment]
Thanks for taking a look at it!

**Comment by sbobbo** [comment]
I am running into this pretty badly now. I've tried the nightly but it doesn't seem to help. Was the solution of moving to a separate thread done? 

Anyone else running into it still?

**Comment by sbobbo** [comment]
The current nightly seems to have resolved this, fwiw.

**Comment by sbobbo** [comment]
I am running into this again, likely due to the number of things i am monitoring now.

The issue has a few side effects:

favorites.json's "lastcheck" entries will not be updated between the time of the program launching and crashing.  This is good and bad, but mostly bad. If it has been multiple intervals since the last time the file was saved, then at next launch, many files will attempt to download but will be met with "file already exists. And of course the program will crash again without manual editing of favorites.json, due to the same memory issue.

The only good part of that is that since the monitor's checks per tag goes faster than the downloads do, if favorites.json was saved upon the crash, it would save new timestamps for tags that had not yet had an opportunity to download. Upon the next time the program is launched, it will use these new timestamps and skip running for those other tags, forever skipping any files posted in that timeframe between the last successful download of the tag and the newest timestamp. This can be confirmed by having many tags monitored and then closing the program manually while it has a gap between the current monitor check and the current monitor download.

"While there is always only one monitor running at the same time, if you have multiple monitors set with the same interval,"

Is the same interval the issue, or is it that all of them are trying to start at the same time. My issue here is that when a crash does occur, i normally don't notice for most of a day, so when i relaunch grabber, it is well past my standard interval time of 18000s. Would varying that interval time actually help in the case where every tag is well beyond their respective interval time?

If the solution is for me to simply change all of my intervals to like 18001, 18002, etc, then that's not too big of a problem...

**Comment by Rodric12** [comment]
Has there been a solution to this problem yet?

**Comment by sbobbo** [comment]
No, I am still experiencing it. I use modifications to the favorites.json file as a workaround to get everything downloaded. It also seemed like setting simultaneous downloads to 10 helped a bit, as did increasing the interval time for the tags I am monitoring.

Hopefully a fix for the memory leak is on the road map.

**Comment by sbobbo** [comment]
@Bionus 
I'm excited to see that you are working on the monitoring features! I don't suppose you have had a chance to look into this issue?

**Comment by Bionus** [comment]
I'll have a look.

How many things are you monitoring at the same time, and how often do new images get added? (i.e. how many images must be downloaded each time you open Grabber)

**Comment by sbobbo** [comment]
Hundreds of items are being monitored. I could share my favorites.json file if that would help. I monitor a few types of things, artists, characters, copyright, and then certain more generic tags like "3D Bikini". Most of the tags are artists, but those are the least likely to bring in images each time. Whereas characters, copyrights, & generic tags are much fewer in number, but they generally result in more images per search. I'm fairly sure I was hitting these crashes back when i was only searching for artists.

If i shared the file, it would be very easy to reproduce. You could do a find/replace on the directory structure and then another find/replace to move the date back 24-48 hours(or i can just send an older copy) and then after opening grabber it would crash before monitoring half of them. On next open, it would monitor all the same tags but probably get a little further because it seems to process the "file already exists" event faster than the actual file downloads

**Comment by Bionus** [comment]
I'll gladly take you up on your offer for your `favorites.json` file.

Because while I have a few ideas as to what can cause the crash, there might be multiple causes for said crash so having "real world" data would be more useful than me creating 200 "rating:safe" monitors.

If you're not comfortable sharing the file here, you can always send it to me at bio.nus@hotmail.fr. I'll take a look tonight after I'm off work. As you said, I'd like to have monitors a main feature of the next version, so if it keeps crashing that won't do 😅

**Comment by sbobbo** [comment]
Sent the files, thanks again!

**Comment by Bionus** [comment]
There's indeed lots of issues with Grabber when using the files you sent, thanks 👌 
I've been working on it yesterday and I'll keep working on it to:
* Make the program more responsive
* Stop eating that much RAM

**Comment by sbobbo** [comment]
Awesome. 

Well, let me reinforce something I mentioned above. The saving of monitor status in the case of a crash might be an issue if it mirrors the saving that occurs when the program is normally closed during monitoring. This is particularly noticeable if you change the "simultaneous downloads" setting back down to "1". In my settings.ini, I have it set to 10 due to it helping with getting more downloaded before the crash.

Okay, so what I'm getting at is that the monitor has 2 main steps from what I can tell. Searching and downloading. When "simultaneous downloads" is set to a lower number, it's much more apparent that downloads can lag many entries behind the searches. I've seen instances of downloads still being on a word starting with like "C" when it's searching a "J" letter. In this scenario, manually closing the program results in the favorites.json being updated so that every tag up to and including the latest "J" tag having its lastCheck being set to a current timestamp. This is a problem because the downloading was still back at that "C" tag. All images for the tags between "C" and "J" will never be downloaded, even on future monitor runs.

If saving monitor status during a crash follows that mechanism, the issue could silently happen as well. Just a thought, Maybe it's possible to only update "lastCheck " based on when any download is finished or when it's flagged to not need a download, rather than when the search executes as part of the monitor?

Does that make sense? I hope I explained it well. Thanks again.

**Comment by Bionus** [comment]
Monitors indeed have multiple steps:
1. Perform the search and see if there are new images
2. Add the images to some "need to download" list
3. In the background, the images to download will be taken from the "need to download" list then downloaded (using your "simultaneous downloads" setting)

Therefore, if you have only 1 simultaneous download but hundreds/thousands of monitors, they will all be "checked" and if new images are found, they will all be added to the "need to download" list. On the other hand, downloading images take time so images will be added faster than they are removed from the list. This causes the list to keep increasing in size, until it takes too much memory and crashes the program.

The main fix I worked on is to simply wait for the downloads before starting new monitors. So the case where "J" is being checked wile "C" isn't even finished downloading shouldn't ever happen anymore. It should also solve the saving issue.

The only side effect that I am aware of, is that if you have "simultaneous downloads" set to 10 for example, but each monitor return a single image every time, you will only actually download one image at a time (since new images won't be searched until the current monitor is finished downloading), possibly slightly lowering the overall download speed.

So there are indeed places where there could be improvements, but for now I think stability is the most critical issue.

You can try it out in the nightly version if you want:
https://github.com/Bionus/imgbrd-grabber/releases/nightly

**Comment by sbobbo** [comment]
Amazing improvement! After installing the nightly and running Grabber once, I backdated my LastChecks to 2 days ago for every single tag, started up Grabber again, and it went through my full list without crashing. That never would have worked on 7.2.1!

What's the best way to support development, Patreon or something else? This alone is such a wonderful update.

Now if only monitored tags wouldn't redownload files already grabbed per that tag. It seems like server side changes to tags can prompt redownloads with the way I have my filename configured. And unfortunately, I don't want to change the "if a file already exists globally" setting away from "Save" due to wanting a picture to exist in an artists folder and a characters folder separately. 

Anyway, you're awesome, thanks for tackling this!

**Comment by sbobbo** [comment]
By the way, thank you for also addressing the issue of the search/downloads for the monitors. I like that solution, it's ideal imo. The stated downside is minor compared to the unintended behavior that could be encountered before.

**Comment by Bionus** [comment]
> Amazing improvement! After installing the nightly and running Grabber once, I backdated my LastChecks to 2 days ago for every single tag, started up Grabber again, and it went through my full list without crashing. That never would have worked on 7.2.1!

Glad to hear that! I was worried there'd be additional issues I didn't notice in my tests 😅 

> What's the best way to support development, Patreon or something else? This alone is such a wonderful update.

There's my [Patreon](https://www.patreon.com/bionus) for monthly/recurring donations, or [PayPal](https://www.paypal.me/jvasti) for one-time ones!

> Now if only monitored tags wouldn't redownload files already grabbed per that tag. It seems like server side changes to tags can prompt redownloads with the way I have my filename configured.

I'm not sure I understand properly. You mean your filename contains the image's tags for example, so if someone changes the tags, it will trigger a re-download since the filename changed? This is strange since already checked images should be directly skipped based on their date (compared to the monitor's last check).

Does this happen on every source? Or could it be that it only happens on sources where the "date" information fails to be fetched correctly? In that case it might cause Grabber to think all images are "new" and re-download them (if the filename changed). If that is indeed the reason, a fix could be to update the comparison to use the last checked image's ID and directly stop there next time.

> And unfortunately, I don't want to change the "if a file already exists globally" setting away from "Save" due to wanting a picture to exist in an artists folder and a characters folder separately.

I know it doesn't change anything for the problem you mentioned, but know that there's also a "shortcut" setting to avoid downloading the same image X times in different folders, taking time and hard drive space. Duplicates will be created as shortcuts to the original. The downside being is that if you delete the original you'll have "orphan" shortcuts.

> Anyway, you're awesome, thanks for tackling this!

No problems! I'm glad that my work is being helpful 👌  

> By the way, thank you for also addressing the issue of the search/downloads for the monitors. I like that solution, it's ideal imo. The stated downside is minor compared to the unintended behavior that could be encountered before.

That's also my opinion. Especially given speed is not the critical factor for monitors, while stability is. I'd rather have Grabber slowly run in the background (FYI you can reduce it to the taskbar and leave it there if you mostly use monitors), downloading everything at a steady pace rather than go crazy and crash. 💀

**Comment by sbobbo** [comment]
>  I'm not sure I understand properly. You mean your filename contains the image's tags for example, so if someone changes the tags, it will trigger a re-download since the filename changed? 

Yes, exactly.

> This is strange since already checked images should be directly skipped based on their date (compared to the monitor's last check).

Well, keep in mind that I had just backdated my LastCheck parameters for all of my tags, so it rechecked all files associated with those tags that had been posted over the last 2 days. While I was of course only doing so as a proof of concept at the time, I actually do this on purpose every few weeks. The reason is that many images get added to these sites without tagging and the tagging happens over the following hours and days. Think about it like this. An image by a great artist gets posted at 12:05, but it doesn't have the artist in the tag. My monitor has a tick for the actual artist of that image at 12:10 and doesn't download the image. At 12:15, someone on the website tags that image to the appropriate artist that i am monitoring. The next time the monitor runs, the image does not download because it's checking for posts that happened since 12:10, not 12:05. Separately, because I include %character% and %copyright% in my filenames, if those are changed between my initial download and when i backdate my tags, the images will be redownloaded with new names. 

Obviously, you can only do so much to mitigate a problem like this, and they may not be worth your time. One mitigation that i may go ahead and embrace is significantly increasing the intervals for all of my monitors, but that is only able to cut down on the issue. I can also use separate tools to scan for duplicates. On your end, I guess checking for the MD5 in the folder and renaming the file would be a solid option, but that may be a ton of work.

> There's my Patreon for monthly/recurring donations, or PayPal for one-time ones!

Is paypal better? I would assume that patreon takes more of a cut.

The shortcut solution doesn't work for me, but thank you for mentioning it.

**Comment by Bionus** [comment]
I see two possible solutions for the problem you mention (and both can be used at the same time).

First option would be to have a setting to "delay" monitors, that is they won't check images that came out in the last X minutes/hours/days, giving some leeway for the tags to be updated. One downside being that if tags are updated much later the issue will remain, but it should at least occur much less often. And another being that the images won't be directly downloaded, since there will be a delay before treating the images.

The other is actually very similar to what you suggested:

> On your end, I guess checking for the MD5 in the folder and renaming the file would be a solid option, but that may be a ton of work.

Rather than checking everytime a file is downloaded for all the MD5 in the directory, another option would be to split the "if a file already exists globally" into two: one for the current folder and one for other folders. So in your case, you would use "Copy" for the "other folders" and "Rename" for the same folder. But that wouldn't work if the folder name itself contains the changed tag.

I feel like the first solution makes a lot of sense, while the second might work only in some cases (where the tags are not in the folder name).

> Is paypal better? I would assume that patreon takes more of a cut.

On PayPal there are indeed no fees, while on Patreon there'd be a cut of around 8% (5% Patreon fee + payment fee). But in the end whichever to use mainly depends on whether you want to make it recurring (Patreon) or one-time (PayPal), or also if you simply already have an account in either of them :sweat_smile:

> The shortcut solution doesn't work for me, but thank you for mentioning it.

Yes, I guess using [Windows hardlinks](https://docs.microsoft.com/en-us/windows/win32/fileio/hard-links-and-junctions) would make more sense for most users. It's a kind of shortcut where the data is only stored in one place, and will only be deleted when all "links" to it are deleted. And for the user/other programs, it's totally indistinguishable from a normal file copy.

**Comment by sbobbo** [comment]
> Rather than checking everytime a file is downloaded for all the MD5 in the directory, another option would be to split the "if a file already exists globally" into two: one for the current folder and one for other folders.

I like that, but maybe I am misunderstanding. I only use %search% or %search_1% in the folder name, so if I'm understanding you right, this second option would be a great solution for my setup.

Option #1 is pretty great too. Having it enabled by default would likely confuse some users, but being able to enable it and tweak the timeframe would work. I guess the only thing is that you would have to subtract(add?) the delay to the next monitor tick.

**Comment by Bionus** [comment]
> I like that, but maybe I am misunderstanding. I only use %search% or %search_1% in the folder name, so if I'm understanding you right, this second option would be a great solution for my setup.

In your specific case, it would indeed be a perfect solution. And I think that would be the case for most people as well, which is why I plan on implementing it.

> I guess the only thing is that you would have to subtract(add?) the delay to the next monitor tick.

It's ignoring the images which are less than X hours old, then setting the "last check" to "now - X hours". It shouldn't be very hard to do and would make a lot of sense for the specific issue you pointed out (images which are not properly tagged "yet").

**Comment by Lil-Sassy** [comment]
I noticed that this issue has re-appeared again after the latest update.
